File: leetcode_python/Binary_Search/search-a-2d-matrix.py
# V0 
# IDEA : MATRIX IN ORDER + BRUTE FORCE
# Space: O(1)
# Time:  O(m+n) # worst case
class Solution:
    def searchMatrix(self, matrix, target):
        if len(matrix) == 0:
            return False        
        row, col = 0, len(matrix[0]) - 1
        while row < len(matrix) and col >= 0:
            if matrix[row][col] == target: 
                return True
            elif matrix[row][col] < target: 
                row += 1
            elif matrix[row][col] > target: 
                col -= 1
        return False

# V0'
# IDEA : BINARY SEARCH
# Space: O(1)
# Time:  O(logm + logn)
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix:
            return False

        m, n = len(matrix), len(matrix[0])
        left, right = 0, m * n
        while left < right:
            mid = left + (right - left) / 2
            if matrix[int(mid / n)][int(mid % n)]>= target:
                right = mid
            else:
                left = mid + 1
        return left < m * n and matrix[int(left / n)][int(left % n)] == target

# V1
# https://leetcode.com/problems/search-a-2d-matrix/discuss/351404/Python-Simple-Solution
class Solution:
    def searchMatrix(self, matrix, target):
        if len(matrix) == 0:
            return False

        row, col = 0, len(matrix[0]) - 1
        
        while row < len(matrix) and col >= 0:
            if matrix[row][col] == target: return True
            elif matrix[row][col] < target: row += 1
            elif matrix[row][col] > target: col -= 1
        
        return False

### Test case
s=Solution()
assert s.searchMatrix([[1,2,3],[4,5,6],[7,8,9]], 9) == True
assert s.searchMatrix([[1,2,3],[4,5,6],[7,8,9]], 1) == True
assert s.searchMatrix([[1,2,3],[4,5,6],[7,8,9]], 99) == False
assert s.searchMatrix([[]], 0) == False
assert s.searchMatrix([[]], 100) == False
assert s.searchMatrix([], 100) == False
assert s.searchMatrix([[-1,3,4,-4]], -1) == False
assert s.searchMatrix([[_ for _ in range(3)] for _ in range(4)], -1) == False
assert s.searchMatrix([[_ for _ in range(3)] for _ in range(4)], 2) == True
assert s.searchMatrix([[_ for _ in range(99)] for _ in range(999)], 2) == True

# V1'
# https://leetcode.com/problems/search-a-2d-matrix/discuss/592696/python-super-easy
# IDEA : BRUTE FORCE
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for x in matrix:
            if target in x:
                return True
        return False

# V1''
# https://blog.csdn.net/fuxuemingzhu/article/details/79459314
# https://blog.csdn.net/fuxuemingzhu/article/details/79459200
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix or not matrix[0]:
            return False
        rows = len(matrix)
        cols = len(matrix[0])
        row, col = 0, cols - 1
        while True:
            if row < rows and col >= 0:
                if matrix[row][col] == target:
                    return True
                elif matrix[row][col] < target:
                    row += 1
                else:
                    col -= 1
            else:
                return False

# V1' 
# https://blog.csdn.net/fuxuemingzhu/article/details/79459314
# https://blog.csdn.net/fuxuemingzhu/article/details/79459200
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        return any(target in row for row in matrix)
        
# V2 
# IDEA : BINARY SEARCH
# Space: O(1)
# Time:  O(logm + logn)
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix:
            return False

        m, n = len(matrix), len(matrix[0])
        left, right = 0, m * n
        while left < right:
            mid = left + (right - left) / 2
            if matrix[int(mid / n)][int(mid % n)]>= target:
                right = mid
            else:
                left = mid + 1
        return left < m * n and matrix[int(left / n)][int(left % n)] == target

File: leetcode_java/src/main/java/LeetCodeJava/BinarySearch/SearchA2DMatrix.java
package LeetCodeJava.BinarySearch;

// https://leetcode.com/problems/search-a-2d-matrix/
/**
 * 74. Search a 2D Matrix
 * Solved
 * Medium
 * Topics
 * Companies
 * You are given an m x n integer matrix matrix with the following two properties:
 *
 * Each row is sorted in non-decreasing order.
 * The first integer of each row is greater than the last integer of the previous row.
 * Given an integer target, return true if target is in matrix or false otherwise.
 *
 * You must write a solution in O(log(m * n)) time complexity.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
 * Output: true
 * Example 2:
 *
 *
 * Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
 * Output: false
 *
 *
 * Constraints:
 *
 * m == matrix.length
 * n == matrix[i].length
 * 1 <= m, n <= 100
 * -104 <= matrix[i][j], target <= 104
 *
 */
public class SearchA2DMatrix {

    // V0
    // IDEA : BINARY SEARCH + FLATTEN MATRIX
    /**
     * time = O(log(M * N))
     * space = O(1)
     */
    public boolean searchMatrix(int[][] matrix, int target) {

        int length = matrix[0].length;
        int width = matrix.length;

        // flatten matrix + binary search
        /** NOTE !!! we FLATTEN matrix as 1D array, and do binary search */
        int l = 0;
        int r = length * width - 1;

        while (r >= l){
            int mid = (l + r) / 2;
            /** NOTE !!!
             *   -> whether x or y axis
             *   -> we do with length
             *   -> e.g. y = mid / length
             *   ->      x = mid % length
             */
            int cur = matrix[mid / length][mid % length];
            if (cur == target){
                return true;
            }else if (cur < target){
                l = mid + 1;
            }else{
                r = mid - 1;
            }
        }

        return false;
    }

    // V0-1
    // IDEA: BINARY SEARCH + 2D ARRAY
    /**
     * time = O(M + N)
     * space = O(1)
     */
    public boolean searchMatrix_0_1(int[][] matrix, int target) {

        int l = matrix.length;
        int w = matrix[0].length;

        // edge
        if (matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }
        if (matrix.length == 1 || matrix[0].length == 1) {

            if (matrix.length == 1) {
                for (int x : matrix[0]) {
                    if (x == target) {
                        return true;
                    }
                }
                return false;
            }

            if (matrix[0].length == 1) {
                for (int i = 0; i < l; i++) {
                    if (matrix[i][0] == target) {
                        return true;
                    }
                }
                return false;
            }
        }

        // binary search
        for (int i = 0; i < l; i++) {
            int[] curArray = matrix[i];
            int left = 0;
            int right = curArray.length - 1;
            // do binary search
            if (target >= curArray[left] && target <= curArray[right]) {
                while (right >= left) {
                    int mid = (left + right) / 2;
                    if (curArray[mid] == target) {
                        return true;
                    } else if (curArray[mid] > target) {
                        right = mid - 1;
                    } else {
                        left = mid + 1;
                    }
                }
            }
        }

        return false;
    }

    // V1
    // IDEA : BINARY SEARCH + FLATTEN MATRIX
    // https://leetcode.com/problems/search-a-2d-matrix/editorial/
    /**
     * time = O(log(M * N))
     * space = O(1)
     */
    public boolean searchMatrix_2(int[][] matrix, int target) {
        int m = matrix.length;
        if (m == 0)
            return false;
        int n = matrix[0].length;

        // binary search
        /** NOTE !!! FLATTEN MATRIX */
        int left = 0, right = m * n - 1;
        int pivotIdx, pivotElement;
        while (left <= right) {
            pivotIdx = (left + right) / 2;
            /** NOTE !!! TRICK HERE :
             *
             *   pivotIdx / n : y index
             *   pivotIdx % n : x index
             */
            pivotElement = matrix[pivotIdx / n][pivotIdx % n];
            if (target == pivotElement)
                return true;
            else {
                if (target < pivotElement)
                    right = pivotIdx - 1;
                else
                    left = pivotIdx + 1;
            }
        }
        return false;
    }

}
